Various improvements

toddrjen opened the pull request:
This pull request adds one bug fix and several features.  

First, it fixes a bug where if you start the logger with log file, set `logLevel` to 0, then delete the logger, the file won't be closed.

Second, it turns `logLevel` and `commandWindowLevel` into dependent properties.  This allows validation to be done on them when they are set.  All the methods access the values through real, hidden properties for performance reasons.  This includes making sure the level is valid, allowing specifying the level by name instead of just by number, and not allowing enabling file logging when a file hasn't been opened.

Third, it allows overriding the date format.  This also checks that the new date format is valid before setting it.

toddrjen committed:
Allow setting log levels by name.

toddrjen committed:
Allow setting date format

toddrjen committed:
Fix bug where open files will not be closed if file log level set to 'OFF'.

toddrjen committed:
Don't allow enabling file logging without valid file.

toddrjen committed:
Change logLevel and commandWindowLevel to properties so validation is done on them when set.

dpo commented on +logging/logging.m:
```diff
@@ -19,7 +19,7 @@
       {'normal', 'red', 'green', 'yellow', 'blue', 'brightred'}, ...
       {'%s', '\033[31m%s\033[0m', '\033[32m%s\033[0m', '\033[33m%s\033[0m', ...
-       '\033[34m%s\033[0m', '\033[1;31m%s\033[0m'});
+       '\033[34m%s\033[0m', '\033[1;31m%s\033[0m'});
```
I can't tell what changed on this line.

dpo commented on +logging/logging.m:
```diff
@@ -32,18 +32,34 @@
+  end
+  
+  properties
+      datefmt;
```
Shouldn't `datefmt` have a default value?

dpo commented:
Not entirely sure that 788452d204d003c3d1741e46908de0fec6419208 is the culprit but I'm seeing a slow down in `testspeed.m`.

toddrjen committed:
Set default date format

Set the default date format.  It is set to be the same as the previous hard-coded format.

dpo commented:
`testspeed` gives me

```
1e2 logs when logging only to command window
Elapsed time is 0.130840 seconds.
1e3 logs when logging is off
Elapsed time is 0.105394 seconds.
1e2 logs when logging to file
Elapsed time is 0.048177 seconds.
```

on master, and

```
1e2 logs when logging only to command window
Elapsed time is 0.148633 seconds.
1e3 logs when logging is off
Elapsed time is 0.195424 seconds.
1e2 logs when logging to file
Elapsed time is 0.059833 seconds.
```

with your changes. What do you think is responsible for the difference? The properties?

toddrjen committed:
"Use getters and setters to allow direct property access to datefmt."

toddrjen committed:
"Use integer comparison for validating log levels since Maps are extremely slow."

toddrjen committed:
"Can't take the min or max of a cell"

toddrjen commented:
The problem was actually the `isKey` `Map` method.  It is extremely slow.  I have replaced it with an value-based lookup and now get performance roughly equal to current master.

toddrjen committed:
Let Map provide the error

`Map.isKey` is very slow, so just try to use the level as a key and let the normal error get thrown if need be.

dpo commented:
Many thanks for all that! The timings looks good now. If I may, can I ask a couple more small things:
- add a comment to explain the reason for `getLevelNumber`
- indent with 2 spaces throughout (some places use 4)
- add a short example in the README on changing the date format.

toddrjen committed:
Fix indentation

Set all indentation to be two lines

toddrjen committed:
Add docstring to getLevelNumber

This adds a docstring explaining what `getLevelNumber` does.

toddrjen committed:
Expand README

Add additional methods and properties to the README.

toddrjen commented:
Requested changes have been implemented.

dpo commented:
Many thanks!
